### DISCUSSION TITLE
Help needed: v2rayv5 quic config, strictly following schema, but doesn't pass test

### DISCUSSION THREAD

**Discussion by mega-optimus** [v2fly/v2ray-core#2502] [Q&A]
Hi, 
I'm trying out v5's quic, following the schema in https://github.com/EHfive/v2ray-jsonschema,
**server config** saved as file `config.v5.json`:

```
{
  "log": {
      "access": {
        "type": "File",
        "level": "Warning",
        "path": "/var/log/v2fly/access.log"
    },
    "error": {
        "type": "File",
        "level": "Warning",
        "path": "/var/log/v2fly/error.log"
    }
  },
  
  "inbounds": [{
    "port": 2000,
    "listen": "127.0.0.1",
    "protocol": "vliteu",
    "settings": {
        "password": "redacted",
        "scramblePacket": true,
        "enableFec": true,
        "enableStabilization": true,
        "enableRenegotiation": true,
        "handshakeMaskingPaddingSize": 0
    },
    "streamSettings": {
        "transport": "quic",
        "transportSettings": {
           "key": "abcd",
           "security": {
               "type": "CHACHA20_POLY1305"
           },
           "header": {
           }
        },
        "socketSettings":{
           "mark": 0,
           "tcpFastOpen": false,
           "tproxy": "off",
           "acceptProxyProtocol": false,
           "tcpKeepAliveInterval": 0,
           "tcpKeepAliveIdle": 0,
           "tcpFastOpenQueueLength": 4096,
           "bindToDevice": "",
           "rxBufSize": 0,
           "txBufSize": 0,
           "forceBufSize": false
        }
    },
    "sniffing": {
        "enabled": false,
        "destOverride":  ["http", "tls", "fakedns", "fakedns+others"],
        "metadataOnly": true
    },
    "tag": "vlite"
  }],
  
  "outbounds": [{
    "protocol": "freedom",
    "settings": {},
    "sendThrough": "0.0.0.0",
    "tag": "free"
  },{
    "protocol": "blackhole",
    "settings": {},
    "tag": "blocked"
  }]
}
```

but, when I run `v2ray test  -format=jsonv5  -c  config.v5.json`, the error message is this:
`Test failed: main/commands: failed to load config: [config.v5.json] > infra/conf/v5cfg: unable to build config > infra/conf/v5cfg: unable to load transport config > common/registry: unable to parse json content > Any JSON doesn't have '@type'`

JSON syntax passed validator checking.

Can anybody see why? Thank you!

**Comment by mydogshitgold** [comment]
remove `"header": {}` from `transportSettings`.

**Reply by mega-optimus** [reply]
> remove `"header": {}` from `transportSettings`.

Thanks, your fix seems to solve the problem.

I have two follow-up questions:
1. Should {"key": "abcd"} be set to a GUID? Doesn't its role overlap with {"password": "redacted"} ? (Document has no mention of "key")
2. Quic is using encryption, but why is there no place to specify the certificate file & key file paths?

Thank you!

**Reply by mydogshitgold** [reply]
1. No, in v2ray proxy layer settings (under `settings`) and transport layer settings (under `streamSettings`) are unrelated. `key` is used for QUIC encryption in presence of `transportSettings.security`. 

2. All TLS-related options go to `securitySettings`, which if not present will automatically generate a TLS certificate.

**Reply by mega-optimus** [reply]
> 1. No, in v2ray proxy layer settings (under `settings`) and transport layer settings (under `streamSettings`) are unrelated. `key` is used for QUIC encryption in presence of `transportSettings.security`.
> 2. All TLS-related options go to `securitySettings`, which if not present will automatically generate a TLS certificate.

Thanks ! Very clear.
When I looked at the schema again, I realized that in "github:EHfive_v2ray-jsonschema_v5config:CustomStreamSettings" section, although ["grpc", "tcp", "quic", "kcp", "ws", "tls", "utls"] are listed in parallel, the first 5 are in "transport" category, while the last 2 are in "security" category.

**Reply by mydogshitgold** [reply]
Generated json schemas do not always represent original protobuf messages clearly. the various `config.proto` and `config.go` files in the repo can be used as a clearer reference if you are in doubt. look for messages like `Config` or `SimplifiedConfig`.


e.g. [transport/internet/config.proto](https://github.com/v2fly/v2ray-core/blob/master/transport/internet/config.proto)
